File: macos/Tests/ElevateCoreTests/StepUpTokenCacheTests.swift
import Testing
import Foundation
@testable import ElevateCore

@Suite struct StepUpTokenCacheTests {
    private func token(exp: Date?) -> String {
        var payload: [String: Any] = ["aud": "https://graph.microsoft.com", "acrs": "c10"]
        if let exp { payload["exp"] = Int(exp.timeIntervalSince1970) }
        let body = try! JSONSerialization.data(withJSONObject: payload)
        let b64 = body.base64EncodedString().replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_").trimmingCharacters(in: CharacterSet(charactersIn: "="))
        return "eyJhbGciOiJub25lIn0.\(b64).sig"
    }

    private let scopes = ["https://graph.microsoft.com/User.Read", "https://graph.microsoft.com/RoleManagementPolicy.Read.Directory"]

    @Test func handsBackTheStepUpTokenForTheSameIdentityTenantAndScopes() async {
        let cache = StepUpTokenCache()
        let stepped = token(exp: .now.addingTimeInterval(3600))
        await cache.store(stepped, identityId: "a", tenantId: "t", scopes: scopes)
        #expect(await cache.token(identityId: "a", tenantId: "t", scopes: scopes.reversed()) == stepped)
    }

    @Test func holdsNothingForAnotherIdentityTenantOrScopeSet() async {
        let cache = StepUpTokenCache()
        await cache.store(token(exp: .now.addingTimeInterval(3600)), identityId: "a", tenantId: "t", scopes: scopes)
        #expect(await cache.token(identityId: "b", tenantId: "t", scopes: scopes) == nil)
        #expect(await cache.token(identityId: "a", tenantId: "other", scopes: scopes) == nil)
        #expect(await cache.token(identityId: "a", tenantId: "t", scopes: ["https://management.azure.com/user_impersonation"]) == nil)
    }

    @Test func dropsTheTokenOnceItsOwnExpiryIsNear() async {
        let clock = Clock()
        let cache = StepUpTokenCache(now: { clock.now })
        await cache.store(token(exp: clock.now.addingTimeInterval(600)), identityId: "a", tenantId: "t", scopes: scopes)
        clock.advance(600 - StepUpTokenCache.skew + 1)
        #expect(await cache.token(identityId: "a", tenantId: "t", scopes: scopes) == nil)
    }

    @Test func anOpaqueTokenIsHeldOnlyLongEnoughForTheRetry() async {
        let clock = Clock()
        let cache = StepUpTokenCache(now: { clock.now })
        await cache.store("not-a-jwt", identityId: "a", tenantId: "t", scopes: scopes)
        #expect(await cache.token(identityId: "a", tenantId: "t", scopes: scopes) == "not-a-jwt")
        clock.advance(StepUpTokenCache.opaqueLifetime)
        #expect(await cache.token(identityId: "a", tenantId: "t", scopes: scopes) == nil)
    }

    @Test func signingOutForgetsWhatWasHeldForThatIdentityAlone() async {
        let cache = StepUpTokenCache()
        let kept = token(exp: .now.addingTimeInterval(3600))
        await cache.store(token(exp: .now.addingTimeInterval(3600)), identityId: "a", tenantId: "t", scopes: scopes)
        await cache.store(kept, identityId: "b", tenantId: "t", scopes: scopes)
        await cache.forget(identityId: "a")
        #expect(await cache.token(identityId: "a", tenantId: "t", scopes: scopes) == nil)
        #expect(await cache.token(identityId: "b", tenantId: "t", scopes: scopes) == kept)
    }

    private final class Clock: @unchecked Sendable {
        private let lock = NSLock()
        private var date = Date(timeIntervalSince1970: 1_700_000_000)
        var now: Date { lock.withLock { date } }
        func advance(_ seconds: TimeInterval) { lock.withLock { date += seconds } }
    }
}

File: windows/tests/Elevate.Core.Tests/StepUpTokenCacheTests.cs
using System.Text.Json;
using Elevate.Core.Auth;
using FluentAssertions;

namespace Elevate.Core.Tests;

/// <summary>Port of the Swift <c>StepUpTokenCacheTests</c>.</summary>
public class StepUpTokenCacheTests
{
    private static readonly string[] Scopes =
    [
        "https://graph.microsoft.com/User.Read",
        "https://graph.microsoft.com/RoleManagementPolicy.Read.Directory",
    ];

    private static readonly DateTimeOffset Start = DateTimeOffset.FromUnixTimeSeconds(1_700_000_000);

    [Fact]
    public void HandsBackTheStepUpTokenForTheSameIdentityTenantAndScopes()
    {
        var cache = new StepUpTokenCache();
        var stepped = Token(DateTimeOffset.UtcNow.AddHours(1));

        cache.Store(stepped, "a", "t", Scopes);

        cache.Token("a", "t", [.. Scopes.Reverse()]).Should().Be(stepped);
    }

    [Fact]
    public void HoldsNothingForAnotherIdentityTenantOrScopeSet()
    {
        var cache = new StepUpTokenCache();
        cache.Store(Token(DateTimeOffset.UtcNow.AddHours(1)), "a", "t", Scopes);

        cache.Token("b", "t", Scopes).Should().BeNull();
        cache.Token("a", "other", Scopes).Should().BeNull();
        cache.Token("a", "t", ["https://management.azure.com/user_impersonation"]).Should().BeNull();
    }

    [Fact]
    public void DropsTheTokenOnceItsOwnExpiryIsNear()
    {
        var now = Start;
        var cache = new StepUpTokenCache(() => now);
        cache.Store(Token(Start.AddMinutes(10)), "a", "t", Scopes);

        now = Start.AddMinutes(10) - StepUpTokenCache.Skew;

        cache.Token("a", "t", Scopes).Should().BeNull();
    }

    [Fact]
    public void AnOpaqueTokenIsHeldOnlyLongEnoughForTheRetry()
    {
        var now = Start;
        var cache = new StepUpTokenCache(() => now);
        cache.Store("not-a-jwt", "a", "t", Scopes);

        cache.Token("a", "t", Scopes).Should().Be("not-a-jwt");

        now = Start + StepUpTokenCache.OpaqueLifetime;

        cache.Token("a", "t", Scopes).Should().BeNull();
    }

    [Fact]
    public void SigningOutForgetsWhatWasHeldForThatIdentityAlone()
    {
        var cache = new StepUpTokenCache();
        var kept = Token(DateTimeOffset.UtcNow.AddHours(1));
        cache.Store(Token(DateTimeOffset.UtcNow.AddHours(1)), "a", "t", Scopes);
        cache.Store(kept, "b", "t", Scopes);

        cache.Forget("a");

        cache.Token("a", "t", Scopes).Should().BeNull();
        cache.Token("b", "t", Scopes).Should().Be(kept);
    }

    /// <summary>A JWT-shaped token that expires at the given moment.</summary>
    private static string Token(DateTimeOffset expiry)
    {
        var payload = new Dictionary<string, object>
        {
            ["aud"] = "https://graph.microsoft.com",
            ["acrs"] = "c10",
            ["exp"] = expiry.ToUnixTimeSeconds(),
        };
        var b64 = Convert.ToBase64String(JsonSerializer.SerializeToUtf8Bytes(payload))
            .Replace('+', '-').Replace('/', '_').TrimEnd('=');
        return $"eyJhbGciOiJub25lIn0.{b64}.sig";
    }
}
